escrow/license_contract.proto

```diff
@@ -0,0 +1,184 @@
+    rpc GetAllLicensesForUser (CallerAuthentication) returns (AllLicensesResponse) {
+    }
+    //this is to only tell you the types of licenses supported by provider , we will start with subscription
+    //and then scale this up to Tier Pricing
+    rpc GetLicensesSupportedByProvider (LicenseProviderReadRequest) returns (LicenseTypesSupportedResponse) {
+    }
+
+    //filter by licenses created for given service Id
+    rpc GetAllLicenseByServiceIds (LicenseProviderReadRequest) returns (AllLicensesResponse) {
+    }
+
+    rpc GetLicenseUsage (LicenseReadRequest) returns (UsageResponse) {
+    }
+
+    //To Be called By Daemon ONLY, Daemons's address is listed in service metadata, this will be used by the licensing server
+    //to check for authorizations on Signature.
+    rpc CheckEligibilityAndIncrementUsage (UpdateLicenseUsageRequest) returns (UsageResponse) {
+    }
+
+    //To Be called By Daemon ONLY, Daemons's address is listed in service metadata, this will be used by the licensing server
+    //    //to check for authorizations on Signature.
+    rpc DecrementUsage (UpdateLicenseUsageRequest) returns (UsageResponse) {
+    }
+
+
+}
+
+
+message CallerAuthentication {
+    //Signature will compromise the following data:
+    // (_method name, current block, user address) signed by private key
+    // Example ("_GetLicense", "block_number") OR ("_CreateLicense", "block_number")
+    bytes signature = 1;
+    //current block number (signature will be valid only for short time around this block number)
+    uint64 current_block = 2;
+}
+
+message UpdateLicenseUsageRequest {
+    CallerAuthentication auth = 1;
+    uint64 channel_id = 2;
+    string service_id = 3;
+
+}
+
+message LicenseCreateRequest {
+    CallerAuthentication auth = 1;
+    //the channel exclusively created for this license needs to be passed
+    uint64 channel_id = 2;
+    //you can pass the list of Authorized addresses to also be allowed to use the License.
+    repeated string user_addresses = 4;
+}
+
+message LicenseFilterByRequest {
+    CallerAuthentication auth = 1;
+    enum Operation {
+        service = 0;
+        Renew = 1;
+    }
+
+
+}
+message LicenseReadRequest {
+    CallerAuthentication auth = 1;
+    //the channel exclusively created for this license needs to be passed
+    uint64 channel_id = 2;
+}
+
+//Based on the given  service Ids passed all licenses associated with the user who has signed will be sent back
+message LicenseProviderReadRequest {
+    CallerAuthentication auth = 1;
+    repeated string service_id = 2;
+
+}
+
+message LicenseUpdateRequest {
+    //Caller authentication data
+    CallerAuthentication auth = 1;
+
+    uint64 channel_id = 2;
+
+    enum Operation {
+        Cancel = 0;
+        Renew = 1;
+    }
+    Operation operation = 3;
+
+    string service_id = 4;
+
+    string license_id = 5;
+}
+
+message DiscountData {
+    uint64 percentage_discount = 1;
+    string license_type = 2;
+}
+
+message UsageResponse {
```

Comment:
Could you please document this message more verbose. It is a main message used by clients and I am not sure how to understand fields properly.

Reply:
Will do